backupccl: memory monitor restore

pbardea opened the pull request:
Fixes https://github.com/cockroachdb/cockroach/issues/64936.

**backupccl: split out SST opening and ingestion**
This commit splits out opening remote SSTs and ingesting them as 2
separate phases of RESTORE. This is useful to allow a future change to
dynamically adjust how many SSTs we open at a time depending on memory
availability.

Release note: None

**backupccl: memory monitor restore SST buffering**
This commit grows the SST batcher's memory monitor used by restore so
that it flushes when it runs into memory pressure.

Release note: None


Release justification: high-benefit fix to control memory usage in SQL
layer during RESTORE.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67477)
<!-- Reviewable:end -->

pbardea committed:
backupccl: add memory accumulator

This commit adds a thin wrapper around a mon.BoundAccount that doesn't
release any memory reserved until closing. It is useful for users that
want to ensure that they can continue using memory they previously
reserved during their entire lifecycle.

Release note: None

pbardea committed:
storageccl: return estimated memory usage for ExternalSSTReader

Release note: None

pbardea committed:
backupccl: memory monitor restore processor

This commit adds memory monitoring to the RESTORE processor. It creates
2 accounts, one for memory to be used for opening iterators and another
that accounts for memory used for buffering the data to ingest.

It uses a light wrapper around a BoundAccount which accumulates memory
(which is only released on termination of the processor). It ensures
that once the machinery starts running we should have sufficient memory
to continue to make progress. If there is not enough memory for all
workers to flush, then the workers will shut down. We ensure that there
is enough memory for at least 1 worker by reserving one worker's worth
of memory up front.

Release note (ops change): Memory used by RESTORE is now accounted for
under the bulk memory monitor.

Release justification: high-benefit fix to control memory usage in SQL
layer during RESTORE.

pbardea commented:
@dt This should be RFAL, the first 2 commits here are in https://github.com/cockroachdb/cockroach/pull/68105.

pbardea committed:
backupccl: add restore memory monitor

Introduces a memory monitor that will be used by RESTORE.

Release note: None

pbardea commented:
@dt This should be ready for another look (has been rebased on the other changes). Looking at it again, it would probably benefit from a unit test restoring a backup with incremental layers with a memory restrictive monitor. Will work on adding those tests to this PR; but I think it should be ready for an initial review.

dt commented on pkg/ccl/storageccl/external_sst_reader.go:
```diff
@@ -84,30 +91,33 @@ func ExternalSSTReader(
 	}
 
 	var reader sstable.ReadableFile = raw
+	var memSize int64
```
I'd maybe init this to remoteReaderMemoryOverhead, and then below, in the non-enc branch, add the remoteSSTSuffixCacheSize

pbardea replied:
I re-assigned the memSize in the enc branch since the size of that buffer depends on the format of the encrypted SSTs (either the entire buffer or chunked).

dt replied:
sure, I guess my thinking is if decryptingReader _doesn't_ download the whole file (which it shouldn't), then we still have all the usual remoteReaderMemoryOverhead coming from the reader under it. /shrug

pbardea replied:
Gotcha, updated to add.

adityamaru commented:
Bazel flake seems unrelated.

bors r=dt
